### PR TITLE
Move simulation_time and emulated_time to shadow-shim-helper-rs

### PR DESCRIPTION
**Pull request by sporksmith**
We'll need the corresponding types in other code there.

This also moves the canonical definitions of the related types and constants into
these Rust modules, and uses cbindgen to re-export them for the C code. e.g. cshadow::SimulationTime is now simulation_time::CSimulationTime.

### CODE REVIEW

**Comment by stevenengler** [comment]
> cshadow::SimulationTime is now simulation_time::CSimulationTime.

Why have separate types for C and Rust? What about only `#[repr(transparent)] pub struct SimulationTime(u64)` (and the same for `EmulatedTime`). The bindings then generate as `typedef uint64_t SimulationTime`, and in the rust `extern` functions you can write them as `pub unsafe extern "C" fn asdf(t: SimulationTime) { ... }` and the C bindings will then look like `void asdf(SimulationTime t);`?

**Comment by sporksmith** [comment]
Yeah, I was thinking about that too. It's a little tricky because some `CSimulationTime` has the `SIMTIME_INVALID` state, which we can return from `checked_add` etc. I think we can probably just have `SimulationTime` and `type OptionSimulationTime = Option<SimulationTime>`, though.

Taking a look now at how that would work out...

**Comment by stevenengler** [comment]
> Yeah, I was thinking about that too. It's a little tricky because some `CSimulationTime` has the `SIMTIME_INVALID` state, which we can return from `checked_add` etc. I think we can probably just have `SimulationTime` and `type OptionSimulationTime = Option<SimulationTime>`, though.

Ah I forgot about `SIMTIME_INVALID`. I guess it's easier to handle that special case if you have separate types that you need to convert between. I'm good with either way.

**Review by stevenengler — APPROVED** [review]

File: src/lib/shadow-shim-helper-rs/src/emulated_time.rs [hunk 1]
```diff
@@ -2,33 +2,49 @@
...
+
+/// The number of nanoseconds from the epoch to January 1st, 2000 at 12:00am UTC.
+/// This is used to emulate to applications that we are in a recent time.
+// cbindgen won't do the constant propagation here. We use the static assertion below
+// to ensure this definition is equal to the intended canonical definition.
+pub const EMUTIME_SIMULATION_START: CEmulatedTime = 946684800u64 * 1_000_000_000u64;
+static_assertions::const_assert_eq!(
+    EMUTIME_SIMULATION_START,
+    SIMULATION_START_SEC * simulation_time::SIMTIME_ONE_SECOND
+);
```
[the remark this answers is not in the document]
Comment (sporksmith) [reply src/lib/shadow-shim-helper-rs/src/emulated_time.rs:33]: Nice; done

File: src/lib/shadow-shim-helper-rs/src/simulation_time.rs [hunk 2]
```diff
@@ -12,10 +12,11 @@ This module contains some identically-named constants defined as C macros in
 use super::emulated_time;
 use std::time::Duration;
 
-use crate::cshadow as c;
-
+#[repr(transparent)]
 #[derive(Copy, Clone, Eq, PartialEq, Debug, PartialOrd, Ord)]
-pub struct SimulationTime(c::SimulationTime);
+pub struct SimulationTime(u64);
```
[the remark this answers is not in the document]
Comment (sporksmith) [reply src/lib/shadow-shim-helper-rs/src/simulation_time.rs:17]: Yeah, good catch. That was left over from when I had initially started trying to unify the types.

[the remark this answers is not in the document]
Comment (sporksmith) [reply src/lib/shadow-shim-helper-rs/src/simulation_time.rs:17]: done

**Pull request merged**